ts/src/main/scala/ch/epfl/ts/evaluation/Evaluator.scala

```diff
@@ -0,0 +1,135 @@
+package ch.epfl.ts.evaluation
+
+import scala.language.postfixOps
+import scala.collection.mutable.{Map => MMap, MutableList => MList}
+import scala.concurrent.duration.{DurationInt, DurationLong}
+import scala.concurrent.duration.FiniteDuration
+import akka.actor.{ActorRef, Cancellable}
+import ch.epfl.ts.component.{ComponentRegistration, Component, ComponentRef, StartSignal, StopSignal}
+import ch.epfl.ts.data._
+import ch.epfl.ts.data.Currency._
+
+/**
+  * Evaluates the performance of traders by total returns, volatility, draw down and sharp
+  *
+  * To use this class, redirect all previous connections into and out of the trader to
+  * instances of this class.
+  *
+  * @param trader the reference to the trader component
+  * @param traderId the id of the trader
+  * @param initial the initial seed money
+  * @param currency currency of the initial seed money
+  * @param period the time period to compute performance
+  */
+class Evaluator(trader: ComponentRef, traderId: Long, initial: Double, currency: Currency, period: FiniteDuration) extends Component {
+  // for usage of scheduler
+  import context._
+
+  private var schedule: Cancellable = null
+  private val wallet = MMap[Currency, Double](currency -> initial)
+  private val returnsList = MList[Double]()
+
+  private var lastValue = initial
+  private var maxProfit = 0.0
+  private var maxLoss = 0.0
+
+  def totalReturns: Double = value() / initial
+  def volatility: Double = computeVolatility
+  def drawdown: Double = maxLoss
+  def sharpeRatio: Double = totalReturns / volatility
+
+  /**
+   * Redirects out-going connections to the trader
+   * */
+  override def connect(ar: ActorRef, ct: Class[_], name: String) = {
+    trader.ar ! ComponentRegistration(ar, ct, name)
+  }
+
+  /**
+   * Handles interested messages and forward all messages to the trader
+   */
+  override def receiver = {
+    case t: Transaction if t.buyerId == traderId =>  // buy
+      trader.ar ! t
+      buy(t)
+    case t: Transaction if t.sellerId == traderId =>  // sell
+      trader.ar ! t
+      sell(t)
+    case 'UpdateStatistics =>
+      updateStatistic
+    case m => trader.ar ! m
+  }
+
+  /**
+   *  Returns the exchange ratio between two currency
+   * */
+  private def ratio(from: Currency, to: Currency): Double = ???
```

Comment:
@merlinND I saw you replied to my question about this method in slack. Thanks for the answer, but it seems I didn't state the problem clearly. The problem is not with how to update the wallet when a transaction is done(`sell` and 'buy' are well implemented).

The problem is to calculate the total value of the wallet in a given currency. In my mind, the market simulator should know this. This is also a common drawback for filter & pipe architectures. There are several approaches to the problem:

- the market simulator to broadcast the information & components cache it(we may need a broadcast mechanism for global data)
  - periodical broadcast is good for frequent change
  - broadcast on change is good if the change frequency is low
- components asks the market simulator to query the info, caches it and periodically ask the simulator to update the cache.

In my mind, something like broadcast mechanism is better.